credentials/jwt/jwt_token_file.go

```diff
@@ -0,0 +1,285 @@
+	retryAttempt    int              // Current retry attempt number
+	nextRetryTime   time.Time        // When next retry is allowed
+
+	// Pre-emptive refresh mutex
+	refreshMu sync.Mutex
+}
+
+// NewTokenFileCallCredentials creates PerRPCCredentials that reads JWT tokens
+// from the specified file path.
+//
+// tokenFilePath is the filepath to the JWT token file.
+func NewTokenFileCallCredentials(tokenFilePath string) (credentials.PerRPCCredentials, error) {
+	if tokenFilePath == "" {
+		return nil, fmt.Errorf("tokenFilePath cannot be empty")
+	}
+
+	return &jwtTokenFileCallCreds{
+		tokenFilePath:   tokenFilePath,
+		backoffStrategy: backoff.DefaultExponential,
+	}, nil
+}
+
+// GetRequestMetadata gets the current request metadata, refreshing tokens
+// if required. This implementation follows the PerRPCCredentials interface.
+// The tokens will get automatically refreshed if they are about to expire or if
+// they haven't been loaded successfully yet.
+// If it's not possible to extract a token from the file, UNAVAILABLE is returned.
+// If the token is extracted but invalid, then UNAUTHENTICATED is returned.
+// If errors are encoutered, a backoff is applied before retrying.
+func (c *jwtTokenFileCallCreds) GetRequestMetadata(ctx context.Context, _ ...string) (map[string]string, error) {
+	ri, _ := credentials.RequestInfoFromContext(ctx)
+	if err := credentials.CheckSecurityLevel(ri.AuthInfo, credentials.PrivacyAndIntegrity); err != nil {
+		return nil, fmt.Errorf("unable to transfer JWT token file PerRPCCredentials: %v", err)
+	}
+
+	// this may be delayed if the token needs to be refreshed from file
+	token, err := c.getToken(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return map[string]string{
+		"authorization": "Bearer " + token,
+	}, nil
+}
+
+// RequireTransportSecurity indicates whether the credentials requires
+// transport security.
+func (c *jwtTokenFileCallCreds) RequireTransportSecurity() bool {
+	return true
+}
+
+// getToken returns a valid JWT token, reading from file if necessary.
+// Implements pre-emptive refresh and caches errors with backoff.
+func (c *jwtTokenFileCallCreds) getToken(ctx context.Context) (string, error) {
+	c.mu.RLock()
+
+	if c.isTokenValid() {
+		token := c.cachedToken
+		shouldRefresh := c.needsPreemptiveRefresh()
+		c.mu.RUnlock()
+
+		if shouldRefresh {
+			c.triggerPreemptiveRefresh()
+		}
+		return token, nil
+	}
+
+	// if still within backoff period, return cached error to avoid repeated file reads
+	if c.cachedError != nil && time.Now().Before(c.nextRetryTime) {
+		err := c.cachedError
+		c.mu.RUnlock()
+		return "", err
+	}
+
+	c.mu.RUnlock()
+	// Token is expired or missing or the retry backoff period has expired. So
+	// refresh synchronously.
+	// NOTE: refreshTokenSync itself acquires the write lock
+	return c.refreshTokenSync(ctx, false)
+}
+
+// isTokenValid checks if the cached token is still valid.
+// Caller must hold c.mu.RLock().
```

Comment:
Please consider adding a `Locked` suffix to these method names so that it is obvious that the method needs to called with the lock held.

Yeah, we used to do a lot of what you have done (basically express the constraint in the comment). But of late, we have been trying to the pattern that I've suggested instead and we feel it works better. Thanks.